File: matlab/qflpn_validation.m
function qflpn_validation()
% QFLPN_VALIDATION
%
% Independent MATLAB validation of the 4-qubit QFLPN reference model.
%
% Validates:
%   1. fuzzy -> quantum mapping
%   2. state normalization
%   3. probability conservation
%   4. density matrix Hermiticity
%   5. density matrix trace
%   6. self-fidelity
%   7. numerical error
%
% No Monte Carlo.
% Fixed deterministic reference instance.

clc;

fprintf('\n');
fprintf('============================================================\n');
fprintf('QFLPN MATLAB VALIDATION\n');
fprintf('============================================================\n');

% ------------------------------------------------------------
% Reference instance
% ------------------------------------------------------------

mu = [0.85, 0.90, 0.45, 0.70];

fprintf('\nReference fuzzy memberships:\n');
fprintf('mu = [%.2f %.2f %.2f %.2f]\n', mu);

% ------------------------------------------------------------
% Core model
% ------------------------------------------------------------

[psi, rho, theta] = qflpn_quantum_core(mu);

% ------------------------------------------------------------
% 1. Fuzzy -> quantum mapping
% ------------------------------------------------------------

mapping_errors = zeros(1, 4);

fprintf('\n[1] Fuzzy -> quantum mapping\n');

for k = 1:4

    probability_one = abs( ...
        cos(theta(k)/2) ...
    )^2;

    % The above expression is P(|0>), therefore P(|1>) is:
    probability_one = abs( ...
        sin(theta(k)/2) ...
    )^2;

    mapping_errors(k) = ...
        abs(probability_one - mu(k));

    fprintf( ...
        'mu(%d)=%.6f | P(|1>)=%.12f | error=%.3e\n', ...
        k, ...
        mu(k), ...
        probability_one, ...
        mapping_errors(k));
end

max_mapping_error = max(mapping_errors);

% ------------------------------------------------------------
% 2. State normalization
% ------------------------------------------------------------

state_norm = norm(psi);
normalization_error = abs(state_norm - 1.0);

fprintf('\n[2] State normalization\n');
fprintf('||psi|| = %.15f\n', state_norm);
fprintf('error   = %.3e\n', normalization_error);

% ------------------------------------------------------------
% 3. Probability conservation
% ------------------------------------------------------------

probabilities = abs(psi).^2;
probability_sum = sum(probabilities);
probability_error = abs(probability_sum - 1.0);

fprintf('\n[3] Probability conservation\n');
fprintf('sum(P)  = %.15f\n', probability_sum);
fprintf('error   = %.3e\n', probability_error);

% ------------------------------------------------------------
% 4. Density matrix Hermiticity
% ------------------------------------------------------------

hermiticity_error = norm( ...
    rho - rho', ...
    'fro' ...
);

fprintf('\n[4] Density matrix Hermiticity\n');
fprintf('error   = %.3e\n', hermiticity_error);

% ------------------------------------------------------------
% 5. Density matrix trace
% ------------------------------------------------------------

trace_value = trace(rho);
trace_error = abs(trace_value - 1.0);

fprintf('\n[5] Density matrix trace\n');
fprintf('trace   = %.15f\n', real(trace_value));
fprintf('error   = %.3e\n', trace_error);

% ------------------------------------------------------------
% 6. Self fidelity
% ------------------------------------------------------------

fidelity = abs(psi' * psi)^2;
fidelity_error = abs(fidelity - 1.0);

fprintf('\n[6] Self-fidelity\n');
fprintf('F       = %.15f\n', fidelity);
fprintf('error   = %.3e\n', fidelity_error);

% ------------------------------------------------------------
% Global numerical error
% ------------------------------------------------------------

maximum_error = max([ ...
    max_mapping_error, ...
    normalization_error, ...
    probability_error, ...
    hermiticity_error, ...
    trace_error, ...
    fidelity_error ...
]);

fprintf('\n============================================================\n');
fprintf('VALIDATION SUMMARY\n');
fprintf('============================================================\n');

fprintf( ...
    'Maximum numerical error = %.3e\n', ...
    maximum_error);

if maximum_error <= 1e-12
    fprintf('STATUS = PASS\n');
else
    fprintf('STATUS = REVIEW REQUIRED\n');
end

fprintf('============================================================\n\n');

end

File: P2/tests/qflpn_validation.py
"""
QFLPN Validation
================

Mathematical and numerical validation of the 4-qubit QFLPN reference model.

Checks:
- fuzzy -> quantum mapping
- normalization
- probability conservation
- density matrix properties
- fidelity
- numerical errors

No Monte Carlo.
"""

from __future__ import annotations

import numpy as np

from qflpn_reference import (
    computational_basis_probabilities,
    density_matrix,
    fuzzy_to_angle,
    pure_state_fidelity,
    reference_state,
    ry_state,
)


REFERENCE_MEMBERSHIPS = np.array(
    [0.85, 0.90, 0.45, 0.70],
    dtype=float,
)


def validate_fuzzy_quantum_mapping(
    memberships: np.ndarray,
) -> float:
    errors = []

    print("\n[1] Fuzzy -> quantum mapping")

    for mu in memberships:
        state = ry_state(float(mu))
        probability_one = abs(state[1]) ** 2
        error = abs(probability_one - mu)
        errors.append(error)

        print(
            f"mu={mu:.6f} | "
            f"P(|1>)={probability_one:.12f} | "
            f"error={error:.3e}"
        )

    return float(max(errors))


def validate_state(state: np.ndarray) -> tuple[float, float]:
    norm = np.linalg.norm(state)
    probabilities = computational_basis_probabilities(state)
    probability_sum = float(probabilities.sum())

    norm_error = abs(norm - 1.0)
    probability_error = abs(probability_sum - 1.0)

    print("\n[2] State validation")
    print(f"Norm                 = {norm:.15f}")
    print(f"Normalization error  = {norm_error:.3e}")
    print(f"Probability sum      = {probability_sum:.15f}")
    print(f"Probability error    = {probability_error:.3e}")

    return norm_error, probability_error


def validate_density_matrix(
    state: np.ndarray,
) -> tuple[float, float]:
    rho = density_matrix(state)

    hermiticity_error = np.linalg.norm(
        rho - rho.conj().T,
        ord="fro",
    )

    trace_error = abs(np.trace(rho) - 1.0)

    print("\n[3] Density matrix validation")
    print(f"Shape                = {rho.shape}")
    print(f"Hermiticity error    = {hermiticity_error:.3e}")
    print(f"Trace                 = {np.trace(rho).real:.15f}")
    print(f"Trace error           = {trace_error:.3e}")

    return float(hermiticity_error), float(trace_error)


def validate_fidelity(state: np.ndarray) -> float:
    fidelity = pure_state_fidelity(state, state)
    fidelity_error = abs(fidelity - 1.0)

    print("\n[4] Fidelity validation")
    print(f"Self-fidelity        = {fidelity:.15f}")
    print(f"Fidelity error       = {fidelity_error:.3e}")

    return fidelity_error


def main() -> None:
    print("=" * 60)
    print("QFLPN 4-QUBIT MATHEMATICAL VALIDATION")
    print("=" * 60)

    print("\nReference memberships:")
    print(REFERENCE_MEMBERSHIPS)

    mapping_error = validate_fuzzy_quantum_mapping(
        REFERENCE_MEMBERSHIPS
    )

    state = reference_state(
        REFERENCE_MEMBERSHIPS
    )

    norm_error, probability_error = validate_state(
        state
    )

    hermiticity_error, trace_error = (
        validate_density_matrix(state)
    )

    fidelity_error = validate_fidelity(state)

    max_error = max(
        mapping_error,
        norm_error,
        probability_error,
        hermiticity_error,
        trace_error,
        fidelity_error,
    )

    print("\n" + "=" * 60)
    print("VALIDATION SUMMARY")
    print("=" * 60)

    print(f"Maximum numerical error = {max_error:.3e}")

    if max_error <= 1e-12:
        print("STATUS: PASS")
    else:
        print("STATUS: REVIEW REQUIRED")


if __name__ == "__main__":
    main()
